### PR TITLE
Add Range Proof Verification to ConvertBack

### PR DESCRIPTION
**Pull request by shawnxie999**
### Summary
ConvertBack now verifies that the remaining confidential balance after conversion is non-negative using bulletproof range proofs. This prevents users from converting more tokens than they have in their confidential balance.

### Changes

**New Cryptographic Functions**
- `computeConvertBackRemainder(commitment, amount, out)` - Computes the Pedersen commitment for the remaining balance: `PC_remainder = PC_balance - amount * G`
- `verifyAggregatedBulletproof(proof, commitments, contextHash)` - Verifies bulletproof range proofs for one or more commitments

**ConvertBack Transaction Updates**
- Preflight validates bulletproof length (688 bytes for single)
- Preclaim computes remaining balance commitment and verifies bulletproof
- Proof verification runs for all paths (with/without auditor) to prevent timing attacks

**New Constants**
- `ecSingleBulletproofLength` = 688 bytes
- `ecDoubleBulletproofLength` = 754 bytes

### Security
- Ensures users cannot convert more than their actual confidential balance
- Range proof verifies remainder is in valid range [0, 2^64 - 1]
- Timing attack prevention: all verification paths execute regardless of early failures

**Test Updates**
- Added `getBulletproof()` helper to generate bulletproofs for testing
- Updated `getConvertBackProof()` to include bulletproof in ZKProof
- Added preflight tests for invalid bulletproof lengths
- Added preclaim tests for invalid/bad bulletproofs (`tecBAD_PROOF`)
- Added edge case tests for minimum amount (1) and maximum amount (maxMPTokenAmount - 1)

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

### CODE REVIEW

File: src/xrpld/app/tx/detail/ConfidentialMPTConvertBack.cpp [hunk 1]
```diff
@@ -70,12 +92,11 @@ verifyProofs(STTx const& tx, std::shared_ptr<SLE const> const& issuance, std::sh
...
         !isTesSuccess(ter))
     {
-        return ter;
+        valid = false;
     }
 
     // Use a pointer to parse each proof component
-    Buffer zkps = Buffer(tx[sfZKProof].data(), tx[sfZKProof].size());
-    std::uint8_t* ptr = zkps.data();
+    std::uint8_t const* ptr = reinterpret_cast<std::uint8_t const*>(tx[sfZKProof].data());
```
Comment (yinyiqian1) [inline src/xrpld/app/tx/detail/ConfidentialMPTConvertBack.cpp]: Let's use  `size_t currentOffset = 0;` instead of a raw pointer.
and retrieve proof with: `proof.substr(currentOffset, somesize);`

Slice class has the `substr` implementation. it is safer and cleaner to use an index

File: src/xrpld/app/tx/detail/ConfidentialMPTConvertBack.cpp [hunk 2]
```diff
@@ -88,13 +109,38 @@ verifyProofs(STTx const& tx, std::shared_ptr<SLE const> const& issuance, std::sh
...
+    // verify bullet proof
+    {
+        // Compute PC_rem = PC_balance - mG (the commitment to the remaining balance)
+        Buffer pcRem;
+        if (auto const ter = computeConvertBackRemainder(tx[sfBalanceCommitment], amount, pcRem); !isTesSuccess(ter))
+        {
+            valid = false;
+        }
+
+        Slice const bulletproofSlice{ptr, ecSingleBulletproofLength};
```
Comment (yinyiqian1) [inline src/xrpld/app/tx/detail/ConfidentialMPTConvertBack.cpp]: For safety, we need to do boundary check every time we retrieve the sub-slice.
Based on my other comment, suggesting using a index.
We can also maintain an variable:
`size_t remainingLength = proof.size();`
And we update remainingLength accordingly, and check
boundary by: 
```
 if (remainingLength < someLength)
        return tecINTERNAL;  // LCOV_EXCL_LINE
```

File: src/test/jtx/impl/mpt.cpp [hunk 3]
```diff
@@ -1859,6 +1862,57 @@ MPTTester::getBalanceLinkageProof(
...
+            contextHash.data()) != 1)
+    {
+        Throw<std::runtime_error>("Bulletproof generation failed");
+    }
+
+    std::size_t const expectedLen = (m == 1) ? ecSingleBulletproofLength : ecDoubleBulletproofLength;
+    if (proofLen != expectedLen)
+        Throw<std::runtime_error>("Unexpected bulletproof length");
+
+    return Buffer(bulletproof.data(), proofLen);
```
Comment (yinyiqian1) [inline src/test/jtx/impl/mpt.cpp:1913]: We can resize the bulletproof buffer to proofLen  and do a std::move to avoid copy.
Buffer class will do a deep copy
```
    Buffer(void const* data, std::size_t size) : Buffer(size)
    {
        if (size)
            std::memcpy(p_.get(), data, size);
    }
```

(Although this will be encapsulated in the mpt-crypto utility layer, let's still change it for now)

File: src/libxrpl/protocol/ConfidentialTransfer.cpp [hunk 4]
```diff
@@ -504,4 +510,119 @@ verifyBalancePcmLinkage(
...
 
+TER
+verifyAggregatedBulletproof(
+    Slice const& proof,
+    std::vector<Slice> const& compressedCommitments,
+    uint256 const& contextHash)
+{
+    // 1. Validate Aggregation Factor (m)
+    std::size_t const m = compressedCommitments.size();
+    if (m == 0)
```
Comment (yinyiqian1) [inline src/libxrpl/protocol/ConfidentialTransfer.cpp]: let's make it cleaner:

`if (m == 0 || (m & (m - 1)) != 0)`

File: src/libxrpl/protocol/ConfidentialTransfer.cpp [hunk 5]
```diff
@@ -504,4 +510,119 @@ verifyBalancePcmLinkage(
...
+        return tecINTERNAL;  // LCOV_EXCL_LINE
+
+    // Bulletproofs require m to be a power of 2
+    if ((m & (m - 1)) != 0)
+        return tecINTERNAL;  // LCOV_EXCL_LINE
+
+    // 2. Prepare Pedersen Commitments, parse from compressed format
+    std::vector<secp256k1_pubkey> commitments(m);
```
Comment (yinyiqian1) [inline src/libxrpl/protocol/ConfidentialTransfer.cpp:528]: let's  call `auto const ctx = secp256k1Context();`, and then use ctx everywhere in this function

File: src/libxrpl/protocol/ConfidentialTransfer.cpp [hunk 6]
```diff
@@ -504,4 +510,119 @@ verifyBalancePcmLinkage(
...
+    if (result != 1)
+        return tecBAD_PROOF;
+
+    return tesSUCCESS;
+}
+
+TER
+computeConvertBackRemainder(Slice const& commitment, std::uint64_t amount, Buffer& out)
+{
+    if (commitment.size() != ecPedersenCommitmentLength)
```
Comment (yinyiqian1) [inline src/libxrpl/protocol/ConfidentialTransfer.cpp]: Le's make it cleaner:

`if (commitment.size() != ecPedersenCommitmentLength || amount == 0)`

File: src/libxrpl/protocol/ConfidentialTransfer.cpp [hunk 7]
```diff
@@ -504,4 +510,119 @@ verifyBalancePcmLinkage(
...
+
+    auto const ctx = secp256k1Context();
+
+    // Parse commitment from compressed format
+    secp256k1_pubkey pcBalance;
+    if (secp256k1_ec_pubkey_parse(ctx, &pcBalance, commitment.data(), ecPedersenCommitmentLength) != 1)
+        return tecINTERNAL;  // LCOV_EXCL_LINE
+
+    // Convert amount to 32-byte big-endian scalar
+    unsigned char mScalar[32] = {0};
```
Comment (yinyiqian1) [inline src/libxrpl/protocol/ConfidentialTransfer.cpp:596]: we can use `boost::endian`, I find this is used in other places as well

File: src/test/jtx/impl/mpt.cpp (hunk 3 above)
Comment (shawnxie999) [reply src/test/jtx/impl/mpt.cpp:1913]: Unfortunately, the  `Buffer` class doesn't currently support resizing that preserves data. Since the underlying memory is managed by a  `unique_ptr`, we can't move construct to a different size—move semantics transfer ownership of the entire allocation, not a portion of it. Even if we added a `resize()` method, it would only update the logical size while the underlying heap allocation remains the original memory size (again, because of unique ptr). True memory shrinking would require reallocation and copying, defeating the purpose.

That said, this function is only used in tests, so such optimizations are likely unimportant IMO.

**Review by yinyiqian1 — APPROVED** [review]
